### PR TITLE
Fix shader node inputs for Blender 4.0

### PR DESCRIPTION
**Pull request by LKAMinco**
Fix to support materials export for Blender 4.0 with support for older versions.

Blender 4.0 change in Shader input naming:
Specular -> Specular IOR Level
Emission -> Emission Color

Fix Emissive color export since new blender has by default Emissive color 1 1 1 1. Now is emissive color exported when strength has non zero value

### CODE REVIEW

File: addon/i3dio/utility.py [hunk 1]
```diff
@@ -82,10 +82,17 @@ def as_fs_relative_path(filepath: str):
...
 This function implements the same ordering as per https://github.com/blender/blender/blob/b0e7a6db56caf6669b6fade1622710d70b96483e/source/blender/blenlib/intern/string.c#L727,
 with the use of a regex as detailed in this answer on stackoverflow https://stackoverflow.com/a/16090640
 """
+
+
 def sort_blender_objects_by_outliner_ordering(objects: List[BlenderObject]) -> List[BlenderObject]:
     return sorted(objects, key=lambda s: [int(t) if t.isdigit() else t.lower() for t in re.split('(\d+)', s.name)])
+
+
+def blender_version() -> int:
```
Comment (StjerneIdioten) [inline addon/i3dio/utility.py:97]: The name of this function is a bit misleading since you aren't returning the version, but just the version Major. Also why does this need a function to begin with, can't you just use `bpy.app.version` directly?

File: addon/i3dio/node_classes/material.py [hunk 2]
```diff
@@ -144,9 +163,16 @@ def _emissive_from_nodes(self, node):
                     return
             self.logger.debug("Has no Emissivemap")
         r, g, b, a = emission_c
-        if (0, 0, 0, 1) != (r, g, b, a):
-            self.logger.debug("Write emissiveColor")
-            self._write_emission(emission_c)
+
+        if blender_version >= 4:
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:167]:
```suggestion
        if bpy.app.version >= (4, 0, 0)
```
I believe this would be better since the other function you made uses bpy anyway and Blender version is always a 3-value tuple

File: addon/i3dio/node_classes/shape.py [hunk 3]
```diff
@@ -206,11 +206,12 @@ def process_subset(self, mesh, material_name: str, triangle_offset: int = 0):
...
                 blender_vertex = mesh.vertices[mesh.loops[loop_index].vertex_index]
 
+
                 # Add vertex color
                 vertex_color = None
                 if len(mesh.vertex_colors):
-                    # Get the color from the active layer, since only one vertex color layer is supported in GE
-                    vertex_color = mesh.vertex_colors.active.data[loop_index].color
+                    # Get the color from the first layer, since only one vertex color layer is supported in GE
+                    vertex_color = mesh.vertex_colors[0].data[loop_index].color
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/shape.py:214]: What is the reason for this change? Has the API changed?

**Review by StjerneIdioten — CHANGES_REQUESTED** [review]

File: addon/i3dio/node_classes/material.py (hunk 2 above)
Comment (LKAMinco) [reply addon/i3dio/node_classes/material.py:167]: This would check just for 4.0 not for 4.1 and newer versions

Comment (StjerneIdioten) [reply addon/i3dio/node_classes/material.py:167]: It is doing "Bigger than" and tuples are compared lexographically element-by-element.

```python
version = (4, 5, 2)

if version >= (4, 0, 0):
    print(f"{version} is higher/equal than/to 4.0.0")
if version >= (4, 3, 1):
    print(f"{version} is higher/equal than/to 4.3.1")
if version >= (4, 5, 1):
    print(f"{version} is higher/equal than/to 4.5.1")
if not version >= (4, 7, 2):
    print(f"{version} is lower/equal than/to 4.7.2")
```

```shell
(4, 5, 2) is higher/equal than/to 4.0.0
(4, 5, 2) is higher/equal than/to 4.3.1
(4, 5, 2) is higher/equal than/to 4.5.1
(4, 5, 2) is lower/equal than/to 4.7.2 
```

Comment (LKAMinco) [reply addon/i3dio/node_classes/material.py:167]: Yes, I forgot about that, in that case, it can be implemented as u have written

File: addon/i3dio/node_classes/shape.py (hunk 3 above)
Comment (LKAMinco) [reply addon/i3dio/node_classes/shape.py:214]: I have got an error where active returned None, but the model had color attribute. Maybe it would be better idea then using first layer, just check whether active is not None

**Comment by StjerneIdioten** [comment]
Could you remove the changes to `addon/i3dio/utility.py` as well given that you aren't using the function anymore? 😊

File: addon/i3dio/node_classes/material.py [hunk 4]
```diff
@@ -144,9 +163,16 @@ def _emissive_from_nodes(self, node):
...
-            self.logger.debug("Write emissiveColor")
-            self._write_emission(emission_c)
+
+        if bpy.app.version >= (4, 0, 0)
+            has_emission = node.inputs['Emission Strength'].default_value == 0.0
+            if not has_emission:
+                self.logger.debug("Write emissiveColor")
+                self._write_emission(emission_c)
+        else:
+            if (0, 0, 0, 1) != (r, g, b, a):
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:173]:
```suggestion
        elif (0, 0, 0, 1) != (r, g, b, a):

```

File: addon/i3dio/node_classes/material.py [hunk 5]
```diff
@@ -76,8 +76,14 @@ def _resolve_with_nodes(self):
             self.logger.debug(f"Has no Glossmap")
 
     def _specular_from_nodes(self, node):
+        # Blender 4.0 have different input names
+        if utility.blender_version() >= 4:
+            specular_input = 'Specular IOR Level'
+        else:
+            specular_input = 'Specular'
+
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:84]: I would suggest deleting this and applying my next comment

File: addon/i3dio/node_classes/material.py [hunk 6]
```diff
@@ -76,8 +76,14 @@ def _resolve_with_nodes(self):
...
     def _specular_from_nodes(self, node):
+        # Blender 4.0 have different input names
+        if utility.blender_version() >= 4:
+            specular_input = 'Specular IOR Level'
+        else:
+            specular_input = 'Specular'
+
         specular = [1.0 - node.inputs['Roughness'].default_value,
-                    node.inputs['Specular'].default_value,
+                    node.inputs[specular_input].default_value,
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:86]:
```suggestion
                    node.inputs['Specular IOR Level' if bpy.app.version >= (4, 0, 0) else 'Specular'].default_value
```

File: addon/i3dio/node_classes/material.py [hunk 7]
```diff
@@ -117,13 +123,26 @@ def _diffuse_from_nodes(self, node):
                     self.xml_elements['Texture'] = xml_i3d.SubElement(self.element, 'Texture')
                     self._write_attribute('fileId', file_id, 'Texture')
         else:
+            # Blender 4.0 have different input names
+            if utility.blender_version() >= 4:
+                emission_input = 'Emission Color'
+            else:
+                emission_input = 'Emission'
+
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:131]: I would suggest deleting this and applying my next comment

File: addon/i3dio/node_classes/material.py [hunk 8]
```diff
@@ -117,13 +123,26 @@ def _diffuse_from_nodes(self, node):
...
         else:
+            # Blender 4.0 have different input names
+            if utility.blender_version() >= 4:
+                emission_input = 'Emission Color'
+            else:
+                emission_input = 'Emission'
+
             # Write the diffuse colors
-            emission_socket = node.inputs['Emission']
+            emission_socket = node.inputs[emission_input]
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:133]:
```suggestion
            emission_socket = node.inputs['Emission Color' if bpy.app.version >= (4, 0, 0) else 'Emission']
```

File: addon/i3dio/node_classes/material.py [hunk 9]
```diff
@@ -117,13 +123,26 @@ def _diffuse_from_nodes(self, node):
...
     def _emissive_from_nodes(self, node):
-        emission_socket = node.inputs['Emission']
+        blender_version = utility.blender_version()
+        # Blender 4.0 have different input names
+        if blender_version >= 4:
+            emission_input = 'Emission Color'
+        else:
+            emission_input = 'Emission'
+
+        emission_socket = node.inputs[emission_input]
```
Comment (StjerneIdioten) [inline addon/i3dio/node_classes/material.py:145]:
```suggestion
        emission_socket = node.inputs['Emission color' if bpy.app.version >= (4, 0, 0) else 'Emission']
```

File: addon/i3dio/utility.py (hunk 1 above)
Comment (StjerneIdioten) [inline addon/i3dio/utility.py:97]: Remove all of the change in this file. This function isn't used anymore and the rest is just formatting stuff.

**Review by StjerneIdioten — CHANGES_REQUESTED** [review]

**Comment by ssnd292** [comment]
@LKAMinco Please test the commits before pushing, use the ammend commit feature or merge the final commit from a dev branch into the merge branch. 

The last commits generated quite some email spam.

**Comment by LKAMinco** [comment]
Ohh yep, sorry :( thats because i have edited it from github not from my computer. I'll get better.

**Comment by StjerneIdioten** [comment]
> @LKAMinco Please test the commits before pushing, use the ammend commit feature or merge the final commit from a dev branch into the merge branch.

You can also create a series of local commits (Don't push) and test your stuff. Then at the end do an interactive rebase and merge everything into one commit before pushing it ;-)

**Comment by ssnd292** [comment]
> > @LKAMinco Please test the commits before pushing, use the ammend commit feature or merge the final commit from a dev branch into the merge branch.
> 
> You can also create a series of local commits (Don't push) and test your stuff. Then at the end do an interactive rebase and merge everything into one commit before pushing it 😉


This is exactly what Ammend Commit already does :D

**Comment by StjerneIdioten** [comment]
> This is exactly what Ammend Commit already does :D

Yes and no. 
Ammend commmit will immediately rewrite your existing commit. So you will be folding in the changes every time you do one. Rebasing at the end you can keep individual commits around. You can also do a mixture depending on how you like to work 😂
